framework/meta/src/version_history.rs

```diff
@@ -1,113 +1,130 @@
 
 /// Indicates where to stop with the upgrades.
-pub const LAST_UPGRADE_VERSION: &str = LAST_VERSION;
+pub const LAST_UPGRADE_VERSION: FrameworkVersion = LAST_VERSION;
 
-pub const LAST_TEMPLATE_VERSION: &str = LAST_VERSION;
+pub const LAST_TEMPLATE_VERSION: FrameworkVersion = LAST_VERSION;
 
+#[rustfmt::skip]
 /// Known versions for the upgrader.
+pub const VERSIONS: &[FrameworkVersion] = framework_versions![
+    0.28.0, 
+    0.29.0, 
+    0.29.2, 
+    0.29.3, 
+    0.30.0, 
+    0.31.0, 
+    0.31.1, 
+    0.32.0, 
+    0.33.0, 
+    0.33.1, 
+    0.34.0, 
+    0.34.1,
+    0.35.0, 
+    0.36.0, 
+    0.36.1, 
+    0.37.0, 
+    0.38.0, 
+    0.39.0, 
+    0.39.1, 
+    0.39.2, 
+    0.39.3, 
+    0.39.4, 
+    0.39.5, 
+    0.39.6,
+    0.39.7, 
+    0.39.8, 
+    0.40.0, 
+    0.40.1, 
+    0.41.0, 
+    0.41.1, 
+    0.41.2, 
+    0.41.3, 
+    0.42.0, 
+    0.43.0, 
+    0.43.1, 
+    0.43.2,
+    0.43.3, 
+    0.43.4, 
+    0.43.5, 
+    0.44.0, 
+    0.45.0, 
+    0.45.2,
+    0.46.0,
+    0.46.1,
+];
+
 #[rustfmt::skip]
-pub const VERSIONS: &[&str] = &[
-    "0.28.0",
-    "0.29.0",
-    "0.29.2",
-    "0.29.3",
-    "0.30.0",
-    "0.31.0",
-    "0.31.1",
-    "0.32.0",
-    "0.33.0",
-    "0.33.1",
-    "0.34.0",
-    "0.34.1",
-    "0.35.0",
-    "0.36.0",
-    "0.36.1",
-    "0.37.0",
-    "0.38.0",
-    "0.39.0",
-    "0.39.1",
-    "0.39.2",
-    "0.39.3",
-    "0.39.4",
-    "0.39.5",
-    "0.39.6",
-    "0.39.7",
-    "0.39.8",
-    "0.40.0",
-    "0.40.1",
-    "0.41.0",
-    "0.41.1",
-    "0.41.2",
-    "0.41.3",
-    "0.42.0",
-    "0.43.0",
-    "0.43.1",
-    "0.43.2",
-    "0.43.3",
-    "0.43.4",
-    "0.43.5",
-    "0.44.0",
-    "0.45.0",
-    "0.45.2",
-    "0.46.0",
-    "0.46.1",
+pub const CHECK_AFTER_UPGRADE_TO: &[FrameworkVersion] = framework_versions![
+    0.28.0,
+    0.29.0,
+    0.30.0,
+    0.31.0,
+    0.32.0,
+    0.33.0,
+    0.34.0,
+    0.35.0,
+    0.36.0,
+    0.37.0,
+    0.40.0,
+    0.41.0,
+    0.42.0,
+    0.43.0,
+    0.44.0,
+    0.45.2,
+    0.46.0
 ];
 
-/// We started supporting contract templates with version 0.43.0.
-pub fn template_versions() -> &'static [&'static str] {
-    &VERSIONS[33..]
-}
+pub const LOWER_VERSION_WITH_TEMPLATE_TAG: FrameworkVersion = framework_version!(0.43.0);
+pub const LOWER_VERSION_WITH_AUTOGENERATED_JSON: FrameworkVersion = framework_version!(0.44.0);
+pub const LOWER_VERSION_WITH_AUTOGENERATED_WASM: FrameworkVersion = framework_version!(0.45.0);
 
-pub fn validate_template_tag(tag: &str) -> bool {
-    template_versions().iter().any(|&tt| tt == tag)
-}
+/// We started supporting contract templates with version 0.43.0.
+pub fn validate_template_tag(tag_str: &str) -> bool {
+    let tag: FrameworkVersion = FrameworkVersion::from_string_template(tag_str);
 
-pub fn template_versions_with_autogenerated_wasm() -> &'static [&'static str] {
-    &VERSIONS[40..]
+    tag >= LOWER_VERSION_WITH_TEMPLATE_TAG && tag <= LAST_VERSION
```

Comment:
Good catch! We should probably validate all framework versions when parsing.
Then we no longer need conditions like `tag <= LAST_VERSION`.
Maybe add a function like `parse_known_version`